### ISSUE TITLE
Revamped Rating Change Model

### ISSUE THREAD

**Issue by ijyliu** [current12/Stat-222-Project#83]
Create our stepwise progression for the rating change models and implement SMOTE to try to get some predictions that aren't just the majority class all the time.

We are trying to predict: "Upgrade", "Downgrade", "Same Rating".

- Do Logistic Regression first. Copy and adapt the code in [Rating Model - Include Previous Rating](https://github.com/current12/Stat-222-Project/tree/main/Code/Modelling/Logistic%20Regression/Rating%20Model%20-%20Include%20Previous%20Rating) and reference/call the functions in [Logistic_Regression_Functions.py](https://github.com/current12/Stat-222-Project/blob/main/Code/Modelling/Logistic%20Regression/Logistic_Regression_Functions.py). You'll replace the current contents of [Logistic Regression/Change Model](https://github.com/current12/Stat-222-Project/tree/main/Code/Modelling/Logistic%20Regression/Change%20Model). There should be three change models and you should load in the names of the variables for each from here: https://github.com/current12/Stat-222-Project/blob/main/Variable%20Index.xlsx. I just picked some stuff to start with, we should definitely discuss/debate what variables to include.
- We can exclude the class weighting (balanced/none) from the grid search
- After completing this for Logistic Regression, move on to XGBoost

**Comment by ijyliu** [comment]
experiment with balanced forced or not

**Comment by ijyliu** [comment]
Suggestion: implement SMOTE

Run the standard 3 models

Do for both Logistic Regression and XGBoost

**Comment by ijyliu** [comment]
@OwenLin2001 

you should be able to just use the functions in Logistic_Regression_Functions.py without modifying them (except for maybe adding an argument that turns off the class weighting part of the grid search). you will just need to write your own function in the changes folder that does smote. your jupyter notebooks will call functions from Logistic_Regression_Functions.py and this new SMOTE function

in your notebooks, you can create a custom_mapping to use that's like {"downgrade": -1, "same": 0}... etc

**Comment by ijyliu** [comment]
Make hyperparameters grid a function argument.

Make two versions of the three models

- One version has class_weight = 'none' forced and uses SMOTE
- Other version has class_weight = 'balanced' forced, does not use SMOTE

**Comment by ijyliu** [comment]
results: logistic change model

![image](https://github.com/current12/Stat-222-Project/assets/67396209/dd01d731-6333-4d8e-83c8-6f978750c716)

![image](https://github.com/current12/Stat-222-Project/assets/67396209/e46c4225-508e-4362-914c-7853543e3116)

i suggest we drop/stop doing the balanced class weight versions, as they are generating weird results. We should increase the aggressiveness of SMOTE (oversample minority classes more/do something to tradeoff more accuracy in order to get more predicitions of minority classes).
